### ISSUE TITLE
Tracking issue for alloc_layout_extra

### ISSUE THREAD

**Issue by Amanieu** [rust-lang/rust#55724]
Feature gate: `#![feature(alloc_layout_extra)]`

This issue tracks additional methods on `Layout` which allow layouts to be composed to build complex layouts.

### Public API

```rust
pub const fn padding_needed_for(&self, align: usize) -> usize;
pub const fn repeat(&self, n: usize) -> Result<(Layout, usize), LayoutErr>;
pub const fn repeat_packed(&self, n: usize) -> Result<Layout, LayoutErr>;
pub const fn extend_packed(&self, next: Layout) -> Result<Layout, LayoutErr>;
pub const fn dangling(&self) -> NonNull<u8>;
```

The main use case is to construct complex allocation layouts for use with the stable global allocator API. For example:
- [`std::collections::HashMap`](https://github.com/rust-lang/rust/blob/master/src/libstd/collections/hash/table.rs#L657)
- [`hashbrown`](https://github.com/Amanieu/hashbrown/blob/2f2af1d1e9bea9d7315b5a85d9cf5b5514f532fe/src/raw/mod.rs#L143)
- Not exactly an example of use, but it would be very useful in [`crossbeam-skiplist`](https://github.com/crossbeam-rs/crossbeam-skiplist/blob/master/src/base.rs#L99)

### Unresolved Questions

- One concern is that not many of these methods have been extensively used in practice. In the examples given above, only `extend`, `array` and `align_to` are used, and I expect that these will be the most used in practice. `padding_needed_for` is used in the implementation of `Rc::from_raw` and `Arc::from_raw`, but in theory could be superseded by the offset returned by `extend`.
- Related to the above, `repeat_packed` and `extend_packed` have no examples or doctests.
- Should `repeat` include trailing padding on the last element?
- Should `padding_needed_for` take a `T` instead of a `usize`?

**Comment by Amanieu** [comment]
In the case of `array<T>` the offset is already available as `size_of::<T>()`. So it doesn't make sense to add a `usize` to the return value. I would rather remove the `usize` from `repeat`.

**Comment by RalfJung** [comment]
FWIW, we actually have explicit tests that allocate with size 8, alignment 16 -- i.e., with a size that is not a multiple of the alignment: `heap::alloc_system_overaligned_request` and `heap::std_heap_overaligned_request`.

**Comment by glandium** [comment]
I stumbled upon `padding_needed_for` and `pad_to_align`, and it seems they are being too cautious.

That is, per https://doc.rust-lang.org/nightly/src/core/alloc.rs.html#95, `layout.size()` is always <= `usize::MAX - (layout.align() - 1)`.

Which means:
- The maximum value `layout.size()` can have is already aligned for `layout.align()` (`layout.align()` being a power of two, `usize::MAX - (layout.align() - 1)` is a multiple of `layout.align()`)
- Incidentally, any value smaller than that maximum value will align at most to that maximum value.

IOW, `pad_to_align` can not return `Err(LayoutErr)`, except for the layout not respecting its invariants, but we shouldn't care about that.

So it seems to me we can drop the `Result` and make the method return a `Layout` directly.

**Comment by TimDiekmann** [comment]
What exactly needs to be done to stabilize this? What questions has to be answered before?

**Comment by p-avital** [comment]
Hello, just wondering if there's any stable alternative to `Layout::array<T>(n)` to allocate arrays manually right now ? This feature is the only block for me to switch my build to stable :(

**Comment by SimonSapin** [comment]
The alternative is `Layout::from_size_align(size_of::<T>().checked_mul(n)?, align_of::<T>())?`

**Comment by mooli** [comment]
I think there might be a bug in the implementation of `repeat()` when it comes to overaligned `Layout`s. It currently computes the padded size and then multiples that by the repeat count. However, this arguably over-pads, since the last array element does not need to be padded.

**Comment by CAD97** [comment]
I'd like to propose stabilization for at least `extend`, `array`, `align_to`, and `pad_to_align`, after #66256 is merged. cc @rust-lang/libs

The longer these go stabilized, the more ad-hoc re-implementations we're going to see. I know myself that I wrote incorrect layout calculation for `#[repr(C)]` structs in edge cases before switching to directly vendoring the standard library's `Layout` manipulation.

## What is stabilized

The functions `Layout::array<T>(n: usize) -> Result<Layout, LayoutErr>`, `Layout::extend(&self, next: Layout) -> Result<(Layout, usize), LayoutErr>`, `Layout::align_to(&self, align: usize) -> Result<Layout, LayoutErr>`, and `Layout::pad_to_align(&self) -> Layout`.

Combined, these functions allow code to construct and manipulate `#[repr(C)]` layouts while letting the standard library handle the awkward edge cases. For example uses, consider [the usage in `hashbrown`](https://github.com/Amanieu/hashbrown/blob/2f2af1d/src/raw/mod.rs#L143) and [this example for the reference](https://github.com/rust-lang/reference/pull/559#issuecomment-511114421):

```rust
/// Returns the layout of a repr(C) struct with the given field layouts in the given order,
/// and the individual field's offsets.
pub fn repr_c(fields: &[Layout]) -> (Layout, Vec<usize>) {
    let mut offsets = Vec::new();
    let mut layout = Layout::from_size_align(0, 0).unwrap();
    for field in fields {
        let (new_layout, offset) = layout.extend(*field).unwrap();
        offsets.push(offset);
        layout = new_layout;
    }
    // Don't forget trailing padding!
    (layout.pad_to_align().unwrap(), offsets)
}
```

**Comment by RalfJung** [comment]
I am still worried about having to do this:
```
    // Don't forget trailing padding!
    (layout.pad_to_align().unwrap(), offsets)
```
I saw experienced Rust compiler devs forget to call this. This is a huge footgun.

The heavy-weight approach to this would be a dedicated type for layouts where this padding has already been done -- letting the type-system track if `pad_to_align` still needs to be called. But that might be overkill.  However, I feel the least we could do is to
* Call this out very clearly in the docs. The current docs even say about `extend` that "The resulting layout will be the same as that of a C struct containing two fields with the layouts of self and next, in that order", which is factually wrong.
* Add a method like your `repr_c` to `Layout`, so that people don't have to write their own loop to turn a list of fields into a layout.

**Comment by CAD97** [comment]
Automatically padding to align or requiring padding to align is probably not desirable. Consider something like Swift's decoupled size/stride; it's not _incorrect_ to have a layout without trailing padding to align, it's just that this is the common default for both `repr(C)` and `repr(Rust)` (any representation where size=stride).

**Comment by RalfJung** [comment]
> Automatically padding to align or requiring padding to align is probably not desirable. Consider something like Swift's decoupled size/stride; it's not incorrect to have a layout without trailing padding to align, it's just that this is the common default for both repr(C) and repr(Rust) (any representation where size=stride).

Ever decoupling this for real in Rust will be near impossible due to lots of unsafe code already assuming size == stride. But yes.

> But I definitely agree that we should fix the extend docs (I'll draft a PR today) and maybe we should add a repr_c fn, though ideally the signature would be approximately fn Layout:::repr_c<const N: usize>(fields: [Layout; N]) -> (Layout, [Layout; N]) rather than requiring allocation for Vec.

Just make it `repr_c(fields: &[Layout])`, no need to allocate. Or `repr_c(fields: impl IntoIter<Item=&Layout>)` if you want to be fancy.

**Comment by CAD97** [comment]
> Just make it `repr_c(fields: &[Layout])`, no need to allocate. Or `repr_c(fields: impl IntoIter<Item=&Layout>)` if you want to be fancy.

Sorry, I had a typo there. ``fn repr_c` wants to provide the offsets of the fields, which requires returning an array or an out parameter for the space. [Here's a potential const-generic impl:](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=eff2b0ed908a9c33d046d563b7a1e6b3)

```rust
pub fn repr_c<const N: usize>(fields: [Layout; N]) -> Result<(Layout, [usize; N]), LayoutErr> {
    unsafe {
        let mut offsets: MaybeUninit<[usize; N]> = MaybeUninit::uninit();
        let mut layout = Layout::from_size_align_unchecked(0, 0);
        let mut offset = offsets.as_mut_ptr().cast::<usize>();
        for field in &fields[..] {
            let (new_layout, this_offset) = layout.extend(*field)?;
            layout = new_layout;
            ptr::write(offset, this_offset);
            offset = offset.offset(1);
        }
        Ok((layout.pad_to_align()?, offsets.assume_init()))
    }
}
```

**Comment by RalfJung** [comment]
Oh I see. Yeah that's kind-of annoying...

**Comment by pnkfelix** [comment]
> (NB: this is unsound as currently implemented as Layout cannot even temporarily have align 0. But it shows effectively what the impl would be.)

Its been ages since I looked at this stuff, but: Can't you use 1 instead of 0 as the initial align?

**Comment by CAD97** [comment]
[This is what I'm using currently, as it avoids the need for `MaybeUninit` wrangling to avoid zero-initializing a small array.](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=79d3da498b8c9f465c52472151dcde7f) But yes, I think it should work fine if you start with align 1.

```rust
pub fn repr_c<const N: usize>(fields: [Layout; N]) -> Result<(Layout, [usize; N]), LayoutErr> {
    let mut offsets: [usize; N] = unsafe { mem::zeroed() }; // replacement for `[0; N]` which doesn't work yet
    let mut layout = fields[0];
    for i in 1..N {
        let (new_layout, this_offset) = layout.extend(fields[i])?;
        layout = new_layout;
        offsets[i] = this_offset;
    }
    Ok((layout.pad_to_align()?, offsets))
}
```

**Comment by CAD97** [comment]
Status update:

#69362 stabilized (in 1.44) `Layout::align_to`, `pad_to_align`, `extend`, and `array`, and #70362 added `Layout::dangling` to this feature gate.

**Comment by Lokathor** [comment]
I'll speak a moment about `repeat` as a potential user of it with a practical example. My use case is as follows: I would like for a bitmap allocation to have H scanlines. Each scanline is W pixels of size P, and each scanline is also optionally over-aligned to 16 or 32.

This can't be done using `array` because no proper type for this exists, so it would be done with something like `repeat` I suppose. In this case, I need the trailing pad bytes of a scanline to also be trailing on the end of the last scanline as well.

However, manually calling `pad_to_align` to pad out the scanline layout before "repeating" that value wouldn't be the end of the world.

**Comment by sunjay** [comment]
Could `repeat` be stabilized? I'm using it to know the layout of various sizes of arrays. With `repeat` I can just store the `Layout` of a single item and compute the layout of each array on the fly with the array's length. I'm also not storing any proper type, so this can't just be done with `array`.

**Comment by scottmcm** [comment]
For `padding_needed_for`, it could be nice to have it take a specific type (see https://github.com/rust-lang/rust/pull/95361/files#diff-02049689ae3bb7ac98af3418ad8fdca219bfa1227cb6cad31afc72bdbae30e27R28) that's always a valid alignment, rather than needing to document that 

> The return value of this function has no meaning if `align` is not a power-of-two.

**Comment by dfreese** [comment]
Curious, are there any major blockers for stabilizing `dangling`?  I've kept an internal implementation of dangling for some aligned buffer code, but it would be great if that could be removed if there are no major concerns that have come up.

**Comment by thomcc** [comment]
Really, all of these are pretty good and useful, not just dangling. I'd be in favor of stabilizing them.

**Comment by RalfJung** [comment]
There is no "dangling" in the summary issue at the top; is that one incomplete?

**Comment by dfreese** [comment]
Thanks for all of the detail and putting that stabilization PR together, @CAD97!

**Comment by scottmcm** [comment]
> padding_needed_for doesn't have any outstanding questions

Well, I put one in https://github.com/rust-lang/rust/issues/55724#issuecomment-1092410896 above

> >  The return value of this function has no meaning if align is not a power-of-two.
>
> but align is required to be a power of two (it's even encoded in the validity requirement, now)

I think you're looking at the wrong thing -- there's an `align: usize` *parameter* so #95361 didn't fix that.

I'd be happy to make a PR to expose that type so it could be used in `padding_needed_for` -- any thoughts where it would best fit?  (It's current `mem::ValidAlign`, but it could be `core::layout::Alignment` or something...)

**Comment by Kixiron** [comment]
Does `Layout::repeat()` have any holdouts that'd keep it from stabilization?

**Comment by ChrisDenton** [comment]
I think at this point `Layout::repeat()` just needs some improved docs, similar to `Layout::extend()`, to mention the importance of `pad_to_align` when e.g. creating `repr(C)` layouts.

cc @CAD97 who last tried to stabilize some of these methods.

**Comment by dtolnay** [comment]
I have updated the API summary at the top of this issue with the current `const` status of all of these functions. They were made const by https://github.com/rust-lang/rust/pull/102207, and a subset of them subsequently reverted to non-const by https://github.com/rust-lang/rust/pull/110393.

| | before #102207 | before #110393 | after #110393 |
|---|:-:|:-:|:-:|
align_to | | const |
pad_to_align | | const | const
padding_needed_for | const | const | const
repeat | | const |
extend | | const |
repeat_packed | | const |
extend_packed | | const |
array | | const | const
dangling | const | const | const

**Comment by rust-rfcbot** [comment]
The final comment period, with a disposition to **merge**, as per the [review above](https://github.com/rust-lang/rust/issues/55724#issuecomment-3403556034), is now **complete**.

As the automated representative of the governance process, I would like to thank the author for their work and everyone else who contributed.

This will be merged soon.
